### DISCUSSION TITLE
MaxListenersExceededWarning: EventEmitter memory leak detected in MeshCentral server logs

### DISCUSSION THREAD

**Discussion by adimate71** [Ylianst/MeshCentral#6545] [Q&A]
Hi,

I’m encountering an issue with the MeshCentral server logs where the following warning appears repeatedly:

(node:1529) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [Socket]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
(node:1529) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [Socket]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit

(node:1529) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [Socket]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(node:1529) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [Socket]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit

I’m concerned it may point to a potential issue with memory management or improper event listener handling. Additionally, I’m worried that this could indicate a deeper problem in the event-driven architecture that might expose vulnerabilities, such as memory leaks or unhandled errors that could be exploited.

Could you please help me understand the cause of this warning and recommend steps to resolve it?

"meshVersion": "v1.1.33",
    "nodeVersion": "v20.18.0",
    "runMode": "WAN mode",
    "productionMode": true,
    "database": "NeDB",
    "platform": "linux",
    "arch": "x64

Thank you

**Comment by adimate71** [comment]
Could you please advise?

**Reply by si458** [reply]
I haven't been able to replicate this issue at all, the only thing I can suggest is restart the meshcentral service and that will free up memory and let it start again

**Reply by adimate71** [reply]
Hi,
I restarted the service and rebooted the virtual machine. However, the same logs continue to appear in the server’s logs.
Please advise.
Thank you

**Reply by si458** [reply]
@adimate71 again, I can not replicate the issue at the moment, but the minute I can, I can try and fix it

Unless u can explain you did X, then logs appeared, 
Or
I did Y X number of times, and then it appeared

Does it just appear straight away after you start meshcentral or gradually?

**Reply by adimate71** [reply]
I have not made any changes or updated the server (Ubuntu 22.04). 
The logs appear after a few hours after the server restart.
Thank you

**Reply by uni-john** [reply]
@si458 
I dont know if this is still an ongoing issue for this particular person but my google searching led me here.
We receive the same error, but only vary rarley, once every couple of months, so its hard to trace ...
Also apologies if this has already been addressed or no longer relevant.

I THINK the the warning is coming from webserver.js:6710, where a handler is added to the req.socket.on('close', …) to finalize the HTTP traffic accounting.

In Node’s HTTP keep‑alive flow the socket is kept open across many requests, but each request adds a handful of built-in 'close' listeners (HTTP parser, keep-alive watchdog, compression cleanup, WebSocket upgrade bookkeeping, etc.). By the time our code reaches if (typeof req.socket.xbytesRead != 'number') … the socket is already carrying ten listeners, so registering one more pushes it over the default cap and Node shouts about a potential leak even though we only add it once.

Some thoughts about fixes
  - Move the accounting callback off the shared socket and onto a per-request lifecycle event. For example, record the running totals at the beginning of the request and then use res.once('finish', …) (or req.once('close', …) for aborted uploads) to do the final delta math. That way the listener’s lifecycle matches a single request, and we never touch the socket’s listener count. 

Something like the below maybe? This keeps the same accounting logic while avoiding additional close listeners on the shared socket, eliminating the MaxListenersExceededWarning.

```
      if (req.headers.upgrade == 'websocket') {
          obj.trafficStats.httpWebSocketCount++;
      } else {
          obj.trafficStats.httpRequestCount++;
          const socket = req.socket;
          const startBytesRead = socket.bytesRead;
          const startBytesWritten = socket.bytesWritten;
          let accounted = false;

          const finalizeHttpAccounting = function () {
              if (accounted === true) return;
              accounted = true;
              obj.trafficStats.httpIn += (socket.bytesRead - startBytesRead);
              obj.trafficStats.httpOut += (socket.bytesWritten - startBytesWritten);
          };

          res.once('finish', finalizeHttpAccounting); // normal completion
          req.once('close', finalizeHttpAccounting);   // aborted upload/keep-alive reuse
      }
```

  - If you need stay on the socket, bump the limit right before attaching the handler. This silences the warning, but it also masks future leaks on that emitter, so I’d treat it as a stopgap at best.

 ```
    if (req.socket.getMaxListeners && req.socket.listenerCount('close') >= req.socket.getMaxListeners()) {
        req.socket.setMaxListeners(req.socket.getMaxListeners() + 1);
    }
    req.socket.once('close', onCloseForAccounting);
```
I dont know if that's necessarily the issue, but perhaps it will guide you to a possible fix or culprit of the reports?

**Comment by adimate71** [comment]
Hi,

I’m still receiving this message occasionally, and I’m not sure what the best remediation would be.
@si458, could you please advise?

Thank you

**Reply by si458** [reply]
ive not been able to track this down yet im afraid, but its only mearly a warning⚠️ 
so nothing to be alarmed about, ill try have a look when i get chance with things
